File: app/src/main/java/com/jpcn/chatapp/ui/activities/RegisterActivity.java
package com.jpcn.chatapp.ui.activities;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.jpcn.chatapp.R;
import com.rengwuxian.materialedittext.MaterialEditText;

import java.util.HashMap;

public class RegisterActivity extends AppCompatActivity {

    MaterialEditText username;
    MaterialEditText password;
    MaterialEditText mail;
    Button buttonRegister;

    FirebaseAuth auth;
    FirebaseDatabase database;
    DatabaseReference reference;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        Toolbar toolbar = findViewById(R.id.header);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle("Register");
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);


        username = findViewById(R.id.username);
        password = findViewById(R.id.password);
        mail = findViewById(R.id.mail);
        buttonRegister = findViewById(R.id.buttonRegister);

        auth = FirebaseAuth.getInstance();

        buttonRegister.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String userNameValue = username.getText().toString();
                String emailValue = mail.getText().toString();
                String passwordValue = password.getText().toString();

                if (TextUtils.isEmpty(userNameValue) || TextUtils.isEmpty(emailValue) || TextUtils.isEmpty(passwordValue)) {
                    Toast.makeText(RegisterActivity.this, "Todos os campos devem ser preenchidos", Toast.LENGTH_SHORT).show();
                } else if (passwordValue.length() < 6) {
                    Toast.makeText(RegisterActivity.this, "O password deve conter pelo menos 6 caracteres", Toast.LENGTH_SHORT).show();
                } else {
                    register(userNameValue, emailValue, passwordValue);
                }

            }
        });
    }

    private void register(final String username, String mail, String password) {
        auth.createUserWithEmailAndPassword(mail, password)
                .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            FirebaseUser firebaseUser = auth.getCurrentUser();
                            String userId = firebaseUser.getUid();

                            database = FirebaseDatabase.getInstance();
                            reference = database.getReference("users").child(userId);


                            HashMap<String, String> hashMap = new HashMap<>();

                            hashMap.put("id", userId);
                            hashMap.put("username", username);
                            hashMap.put("imageURL", "default");

                            reference.setValue(hashMap).addOnCompleteListener(new OnCompleteListener<Void>() {
                                @Override
                                public void onComplete(@NonNull Task<Void> task) {
                                    if (task.isSuccessful()) {
                                        Intent intent = new Intent(RegisterActivity.this, MainActivity.class);
                                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
                                        startActivity(intent);
                                        finish();
                                    }
                                }
                            });
                        } else {
                            Toast.makeText(RegisterActivity.this, "Não foi possível realizar o cadastro", Toast.LENGTH_LONG).show();
                        }
                    }
                });
    }
}

File: app/src/main/java/com/jpcn/chatapp/ui/activities/RegisterActivity.kt
package com.jpcn.chatapp.ui.activities

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import com.jpcn.chatapp.R
import com.jpcn.chatapp.databinding.ActivityRegisterBinding

class RegisterActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityRegisterBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setupToolbar(binding)

        binding.buttonRegister.setOnClickListener {
            validateRegisterFields(binding)
        }
    }

    private fun validateRegisterFields(binding: ActivityRegisterBinding) {
        val userNameValue = binding.username.text.toString()
        val emailValue = binding.mail.text.toString()
        val passwordValue = binding.password.text.toString()

        if (userNameValue.isBlank() || emailValue.isBlank() || passwordValue.isBlank()) {
            Toast.makeText(applicationContext, getString(R.string.register_error_message), Toast.LENGTH_SHORT).show()
        } else if (passwordValue.length < 6) {
            Toast.makeText(applicationContext, getString(R.string.passoword_error_msg), Toast.LENGTH_SHORT).show()
        } else {
            register(userNameValue, emailValue, passwordValue)
        }
    }

    private fun setupToolbar(binding: ActivityRegisterBinding) {
        setSupportActionBar(binding.header.toolbar)
        supportActionBar?.title = "Register"
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun register(username: String, mail: String, password: String) {
        val auth = FirebaseAuth.getInstance()
        auth.createUserWithEmailAndPassword(mail, password).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                persistUser(auth, username)
            } else {
                Toast.makeText(applicationContext, getString(R.string.register_failed_message), Toast.LENGTH_LONG)
                    .show()
            }
        }
    }

    private fun persistUser(auth: FirebaseAuth, username: String) {
        val firebaseUser = auth.currentUser
        val userId = firebaseUser!!.uid
        val database = FirebaseDatabase.getInstance()
        val reference = database.getReference("users").child(userId)
        val hashMap = hashMapOf(
            "id" to userId,
            "username" to username,
            "imageURL" to "default"
        )
        reference.setValue(hashMap).addOnCompleteListener { task2 ->
            if (task2.isSuccessful) {
                val intent = Intent(this@RegisterActivity, MainActivity::class.java)
                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
                startActivity(intent)
                finish()
            }
        }
    }
}